src/sidebar/test/service-url-test.js

```diff
@@ -1,20 +1,166 @@
 'use strict';
 
-var serviceUrl = require('../service-url');
+var proxyquire = require('proxyquire');
 
-describe('serviceUrl', function () {
-  var service;
+/** Return a fake annotationUI object. */
+function fakeAnnotationUI() {
+  var links = null;
+  return {
+    updateLinks: function(newLinks) {
+      links = newLinks;
+    },
+    getState: function() {
+      return {links: links};
+    },
+  };
+}
 
-  beforeEach(function () {
-    service = serviceUrl({serviceUrl: 'https://test.hypothes.is/'});
+function createServiceUrl(linksPromise) {
+  var replaceURLParams = sinon.stub().returns(
+    {url: 'EXPANDED_URL', params: {}}
+  );
+
+  var serviceUrlFactory = proxyquire('../service-url', {
+    './util/url-util': { replaceURLParams: replaceURLParams },
+  });
+
+  var annotationUI = fakeAnnotationUI();
+
+  var features = { flagEnabled: sinon.stub().returns(true) };
+
+  var settings = {};
+
+  var store = {
+    links: sinon.stub().returns(linksPromise),
+  };
+
+  return {
+    annotationUI: annotationUI,
+    store: store,
+    serviceUrl: serviceUrlFactory(annotationUI, features, settings, store),
+    replaceURLParams: replaceURLParams,
+  };
+}
+
+describe('links', function () {
+  context('before the API response has been received', function() {
+    beforeEach(function() {
+      // Create a serviceUrl function with an unresolved Promise that will
+      // never be resolved - it never receives the links from store.links().
+      var parts = createServiceUrl(new Promise(function() {}));
+
+      this.serviceUrl = parts.serviceUrl;
```

Comment:
By not having to work around undesirable `this` binding in promises, this actually made the tests simpler. It allowed me to remove a lot of `var annotationUI = this.annotationUI;` rows from `it()` functions. I'm reminded of the value of avoiding `this` in JavaScript:

```diff
diff --git a/src/sidebar/test/service-url-test.js b/src/sidebar/test/service-url-test.js
index 9c34eb89..de97c1f4 100644
--- a/src/sidebar/test/service-url-test.js
+++ b/src/sidebar/test/service-url-test.js
@@ -40,26 +40,31 @@ function createServiceUrl(linksPromise) {
 
 describe('links', function () {
   context('before the API response has been received', function() {
+    var annotationUI;
+    var replaceURLParams;
+    var serviceUrl;
+    var store;
+
     beforeEach(function() {
       // Create a serviceUrl function with an unresolved Promise that will
       // never be resolved - it never receives the links from store.links().
       var parts = createServiceUrl(new Promise(function() {}));
 
-      this.serviceUrl = parts.serviceUrl;
-      this.store = parts.store;
+      serviceUrl = parts.serviceUrl;
+      store = parts.store;
     });
 
     it('sends one API request for the links at boot time', function() {
-      assert.calledOnce(this.store.links);
-      assert.isTrue(this.store.links.calledWithExactly());
+      assert.calledOnce(store.links);
+      assert.isTrue(store.links.calledWithExactly());
     });
 
     it('returns an empty string for any link', function() {
-      assert.equal(this.serviceUrl('foo'), '');
+      assert.equal(serviceUrl('foo'), '');
     });
 
     it('returns an empty string even if link params are given', function() {
-      assert.equal(this.serviceUrl('foo', {bar: 'bar'}), '');
+      assert.equal(serviceUrl('foo', {bar: 'bar'}), '');
     });
   });
 
@@ -74,33 +79,30 @@ describe('links', function () {
   });
 
   context('after the API response has been received', function() {
+    var linksPromise;
+
     beforeEach(function() {
       // The links Promise that store.links() will return.
-      this.linksPromise = Promise.resolve({
+      linksPromise = Promise.resolve({
         first_link: 'http://example.com/first_page/:foo',
         second_link: 'http://example.com/second_page',
       });
 
-      var parts = createServiceUrl(this.linksPromise);
+      var parts = createServiceUrl(linksPromise);
 
-      this.annotationUI = parts.annotationUI;
-      this.serviceUrl = parts.serviceUrl;
-      this.replaceURLParams = parts.replaceURLParams;
+      annotationUI = parts.annotationUI;
+      serviceUrl = parts.serviceUrl;
+      replaceURLParams = parts.replaceURLParams;
     });
 
     it('updates annotationUI with the real links', function() {
-      var annotationUI = this.annotationUI;
-
-      return this.linksPromise.then(function(links) {
+      return linksPromise.then(function(links) {
         assert.deepEqual(annotationUI.getState(), {links: links});
       });
     });
 
     it('calls replaceURLParams with the path and given params', function() {
-      var serviceUrl = this.serviceUrl;
-      var replaceURLParams = this.replaceURLParams;
-
-      return this.linksPromise.then(function() {
+      return linksPromise.then(function() {
         var params = {foo: 'bar'};
 
         serviceUrl('first_link', params);
@@ -113,10 +115,7 @@ describe('links', function () {
     });
 
     it('passes an empty params object to replaceURLParams if no params are given', function() {
-      var serviceUrl = this.serviceUrl;
-      var replaceURLParams = this.replaceURLParams;
-
-      return this.linksPromise.then(function() {
+      return linksPromise.then(function() {
         serviceUrl('first_link');
 
         assert.calledOnce(replaceURLParams);
@@ -125,9 +124,7 @@ describe('links', function () {
     });
 
     it('returns the expanded URL from replaceURLParams', function() {
-      var serviceUrl = this.serviceUrl;
-
-      return this.linksPromise.then(function() {
+      return linksPromise.then(function() {
         var renderedUrl = serviceUrl('first_link');
 
         assert.equal(renderedUrl, 'EXPANDED_URL');
@@ -135,9 +132,7 @@ describe('links', function () {
     });
 
     it("throws an error if it doesn't have the requested link", function() {
-      var serviceUrl = this.serviceUrl;
-
-      return this.linksPromise.then(function() {
+      return linksPromise.then(function() {
         assert.throws(
           function() { serviceUrl('madeUpLinkName'); },
           Error, 'Unknown link madeUpLinkName');
@@ -145,14 +140,13 @@ describe('links', function () {
     });
 
     it('throws an error if replaceURLParams returns unused params', function() {
-      var serviceUrl = this.serviceUrl;
       var params = {'unused_param_1': 'foo', 'unused_param_2': 'bar'};
-      this.replaceURLParams.returns({
+      replaceURLParams.returns({
         url: 'EXPANDED_URL',
         params: params,
       });
 
-      return this.linksPromise.then(function() {
+      return linksPromise.then(function() {
         assert.throws(
           function() { serviceUrl('first_link', params); },
           Error, 'Unknown link parameters: unused_param_1, unused_param_2');
```